Blockchain/Anchor framework.md

```diff
@@ -0,0 +1,206 @@
+---
+tags: blockchain, solana, anchor
+author: Pham Ngoc Thanh
+---
+
+## Short tutorial to start smart contract development on Solana using Anchor framework
+
+### Basic knowledge you need to have before you start
+- Basic knowledge of Solana network
+- Concepts of: Account, program, instruction,
+- Knowledge of Rust or C++ language
+- Install and launch local network Solana
+
+##### Start learning what Anchor is: Anchor is a framework as well as a set of libraries that make developing applications on the Solana network faster and easier, However, you need to pay attention, the source code of the framework This has not been audited, please be careful when using. In addition, Anchor also provides a number of tools for developers.
+
+- Rust crates and eDSL for writing Solana programs
+- IDL specification
+- TypeScript package for generating clients from IDL
+- CLI command and workspace management for developing complete programs
+
+### Why do we use Anchor ?
+- Easy to use library
+- The ingredients are clear
+- Many utilities for faster build, test and deploy
+- Having IDL generate client on javascript or Golang easy
+
+### An Anchor program consists of 3 main parts:
+1. Program: where the logic of the program is stored
+2. Account struct: where the account struct is defined, marked with #[derive(Accounts)]
+3. declare_id: Declare the address of the program, it uses this id for security checks as well as allowing other programs to access your key field
+
+Like other networks Solana also develops RPC, creating IDLs and client apps from IDL similar to ABI on ethereum.
+
+In the anchor you need to setup workspace for your project, example workspace for devnet
+```
+[provider]
+cluster = "devnet"
+wallet = "~/.config/solana/id.json"
+
+[programs.devnet]
+basic_0 = "2Fcoq3dKjpZiKko9Ho4NZguunJsd9AiGHheF8VHPnk7D"
+
+[scripts]
+test = "yarn run mocha -t 1000000 tests/"
+```
+If you want to deploy testnet, you need to move cluster to testnet.
+View cluster list:
+```
+anchor cluster list
+```
+
+#### Install Anchor and create a program:
+Step 1. Install the library
+- Rust
+```curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
+source $HOME/.cargo/env
+rustup component add rustfmt
+```
+- Solana:
+```
+sh -c "$(curl -sSfL https://releases.solana.com/v1.9.1/install)"
+```
+- Yarn:
+```
+npm install -g yarn
+```
+- Anchor:
+```
+npm i -g @project-serum/anchor-cli
+```
+Check anchor install: anchor --version
+
+Step 2: Create a program
+```
+anchor init <new-project-name>
+```
+Note, you need to choose the appropriate Anchor version for the sample project or you may get an error when building because the old versions are no longer compatible. Let's take the example of creating a program that accepts a key-value pair
+start with declaring the program you declare in the lib.rs . directory
+```
+use anchor_lang::prelude::*;
+
+declare_id!("Fg6PaFpoGXkYsidMpWTK6W2BeZ7FEfcYkg476zPFsLnS");
```

Comment:
Correct me if I'm wrong, as I know this is the default program id which anchor generated when we inti a new project. After we run `anchor build`, it will display a new program id. We need to replace the old program id with the new one. I think it's worth mentioning this step.

Reply:
Yes, you right, should have a note for it, tks